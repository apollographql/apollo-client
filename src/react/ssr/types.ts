export type BatchOptions = {
  debounce?: number;
}
