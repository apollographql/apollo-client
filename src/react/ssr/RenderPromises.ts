import { DocumentNode } from 'graphql';

import { ObservableQuery } from '../../core';
import { QueryDataOptions } from '../types/types';
import { QueryData } from '../data/QueryData';

type QueryInfo = {
  seen: boolean;
  observable: ObservableQuery<any, any> | null;
};

function makeDefaultQueryInfo(): QueryInfo {
  return {
    seen: false,
    observable: null
  };
}

export class RenderPromises {
  // Map from Query component instances to pending fetchData promises.
  private queryPromises = new Map<QueryDataOptions<any, any>, Promise<any>>();

  // Two-layered map from (query document, stringified variables) to QueryInfo
  // objects. These QueryInfo objects are intended to survive through the whole
  // getMarkupFromTree process, whereas specific Query instances do not survive
  // beyond a single call to renderToStaticMarkup.
  private queryInfoTrie = new Map<DocumentNode, Map<string, QueryInfo>>();

  private stopped = false;
  public stop() {
    if (!this.stopped) {
      this.queryPromises.clear();
      this.queryInfoTrie.clear();
      this.stopped = true;
    }
  }

  // Registers the server side rendered observable.
  public registerSSRObservable<TData, TVariables>(
    observable: ObservableQuery<any, TVariables>,
    props: QueryDataOptions<TData, TVariables>
  ) {
    if (this.stopped) return;
    this.lookupQueryInfo(props).observable = observable;
  }

  // Get's the cached observable that matches the SSR Query instances query and variables.
  public getSSRObservable<TData, TVariables>(
    props: QueryDataOptions<TData, TVariables>
  ) {
    return this.lookupQueryInfo(props).observable;
  }

  public addQueryPromise<TData, TVariables>(
    queryInstance: QueryData<TData, TVariables>,
    finish: () => React.ReactNode
  ): React.ReactNode {
    if (!this.stopped) {
      const info = this.lookupQueryInfo(queryInstance.getOptions());
      if (!info.seen) {
        this.queryPromises.set(
          queryInstance.getOptions(),
          new Promise(resolve => {
            resolve(queryInstance.fetchData());
          })
        );
        // Render null to abandon this subtree for this rendering, so that we
        // can wait for the data to arrive.
        return null;
      }
    }
    return finish();
  }

  public hasPromises() {
    return this.queryPromises.size > 0;
  }

  public consumeAndAwaitPromises() {
    const promises: Promise<any>[] = [];
    this.queryPromises.forEach((promise, queryInstance) => {
      // Make sure we never try to call fetchData for this query document and
      // these variables again. Since the queryInstance objects change with
      // every rendering, deduplicating them by query and variables is the
      // best we can do. If a different Query component happens to have the
      // same query document and variables, it will be immediately rendered
      // by calling finish() in addQueryPromise, which could result in the
      // rendering of an unwanted loading state, but that's not nearly as bad
      // as getting stuck in an infinite rendering loop because we kept calling
      // queryInstance.fetchData for the same Query component indefinitely.
      this.lookupQueryInfo(queryInstance).seen = true;
      promises.push(promise);
    });
    this.queryPromises.clear();
    return Promise.all(promises);
  }

  private lookupQueryInfo<TData, TVariables>(
    props: QueryDataOptions<TData, TVariables>
  ): QueryInfo {
    const { queryInfoTrie } = this;
    const { query, variables } = props;
    const varMap = queryInfoTrie.get(query) || new Map<string, QueryInfo>();
    if (!queryInfoTrie.has(query)) queryInfoTrie.set(query, varMap);
    const variablesString = JSON.stringify(variables);
    const info = varMap.get(variablesString) || makeDefaultQueryInfo();
    if (!varMap.has(variablesString)) varMap.set(variablesString, info);
    return info;
  }
}
