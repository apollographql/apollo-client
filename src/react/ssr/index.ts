export { getMarkupFromTree, getDataFromTree } from './getDataFromTree';
export { renderToStringWithData } from './renderToStringWithData';
export { RenderPromises } from './RenderPromises';
