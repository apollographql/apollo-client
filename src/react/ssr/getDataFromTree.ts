import React from 'react';
import { getApolloContext } from '../context';
import { RenderPromises } from './RenderPromises';

export function getDataFromTree(
  tree: React.ReactNode,
  context: { [key: string]: any } = {}
) {
  return getMarkupFromTree({
    tree,
    context,
    // If you need to configure this renderFunction, call getMarkupFromTree
    // directly instead of getDataFromTree.
    renderFunction: require('react-dom/server').renderToStaticMarkup
  });
}

export type GetMarkupFromTreeOptions = {
  tree: React.ReactNode;
  context?: { [key: string]: any };
  renderFunction?: (
    tree: React.ReactElement<any>,
  ) => string | PromiseLike<string>;
};

export function getMarkupFromTree({
  tree,
  context = {},
  // The rendering function is configurable! We use renderToStaticMarkup as
  // the default, because it's a little less expensive than renderToString,
  // and legacy usage of getDataFromTree ignores the return value anyway.
  renderFunction = require('react-dom/server').renderToStaticMarkup
}: GetMarkupFromTreeOptions): Promise<string> {
  const renderPromises = new RenderPromises();

  function process(): Promise<string> {
    // Always re-render from the rootElement, even though it might seem
    // better to render the children of the component responsible for the
    // promise, because it is not possible to reconstruct the full context
    // of the original rendering (including all unknown context provider
    // elements) for a subtree of the original component tree.
    const ApolloContext = getApolloContext();

    return new Promise<string>(resolve => {
      const element = React.createElement(
        ApolloContext.Provider,
        { value: { ...context, renderPromises }},
        tree,
      );
      resolve(renderFunction(element));
    }).then(html => {
      return renderPromises.hasPromises()
        ? renderPromises.consumeAndAwaitPromises().then(process)
        : html;
    });
  }

  return Promise.resolve().then(process);
}
