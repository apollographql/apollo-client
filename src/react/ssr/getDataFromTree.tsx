import { print } from "graphql";
import type * as ReactTypes from "react";
import * as React from "react";
import { renderToStaticMarkup } from "react-dom/server";
import { filter, firstValueFrom } from "rxjs";

import type {
  DocumentNode,
  ObservableQuery,
  OperationVariables,
} from "@apollo/client/core";
import { getApolloContext } from "@apollo/client/react/context";
import { wrapperSymbol } from "@apollo/client/react/internal";
import { canonicalStringify } from "@apollo/client/utilities";

import { useSSRQuery } from "./useSSRQuery.js";

export function getDataFromTree(
  tree: ReactTypes.ReactNode,
  context: { [key: string]: any } = {}
) {
  return getMarkupFromTree({
    tree,
    context,
    // If you need to configure this renderFunction, call getMarkupFromTree
    // directly instead of getDataFromTree.
    renderFunction: renderToStaticMarkup,
  });
}

type GetMarkupFromTreeOptions = {
  tree: ReactTypes.ReactNode;
  context?: { [key: string]: any };
  renderFunction?: (
    tree: ReactTypes.ReactElement<any>
  ) => string | PromiseLike<string>;
};

type ObservableQueryKey = `${string}|${string}`;
function getObservableQueryKey(
  query: DocumentNode,
  variables: Record<string, any> = {}
): ObservableQueryKey {
  const queryKey = print(query);
  const variablesKey = canonicalStringify(variables);
  return `${queryKey}|${variablesKey}`;
}
export interface GetMarkupFromTreeContext {
  getObservableQuery(
    query: DocumentNode,
    variables?: Record<string, any>
  ): ObservableQuery | undefined;
  onCreatedObservableQuery: (
    observable: ObservableQuery,
    query: DocumentNode,
    variables: OperationVariables
  ) => void;
}

export function getMarkupFromTree({
  tree,
  context = {},
  // The rendering function is configurable! We use renderToStaticMarkup as
  // the default, because it's a little less expensive than renderToString,
  // and legacy usage of getDataFromTree ignores the return value anyway.
  renderFunction = renderToStaticMarkup,
}: GetMarkupFromTreeOptions): Promise<string> {
  const availableObservableQueries = new Map<
    ObservableQueryKey,
    ObservableQuery
  >();
  let recentlyCreatedObservableQueries = new Set<ObservableQuery>();
  const internalContext: GetMarkupFromTreeContext = {
    getObservableQuery(query, variables) {
      return availableObservableQueries.get(
        getObservableQueryKey(query, variables)
      );
    },
    onCreatedObservableQuery: (
      observable: ObservableQuery,
      query: DocumentNode,
      variables: OperationVariables
    ) => {
      availableObservableQueries.set(
        getObservableQueryKey(query, variables),
        observable
      );
      if (observable.options.fetchPolicy !== "cache-only") {
        recentlyCreatedObservableQueries.add(observable);
      }
    },
  };

  function process(): Promise<string> {
    // Always re-render from the rootElement, even though it might seem
    // better to render the children of the component responsible for the
    // promise, because it is not possible to reconstruct the full context
    // of the original rendering (including all unknown context provider
    // elements) for a subtree of the original component tree.
    const ApolloContext = getApolloContext();

    return new Promise<string>((resolve) => {
      const element = (
        <ApolloContext.Provider
          value={{
            ...context,
            [wrapperSymbol]: {
              useQuery: () => useSSRQuery.bind(internalContext),
            },
          }}
        >
          {tree}
        </ApolloContext.Provider>
      );
      resolve(renderFunction(element));
    }).then((html) => {
      if (recentlyCreatedObservableQueries.size == 0) {
        return html;
      }
      return Promise.all(
        Array.from(recentlyCreatedObservableQueries).map(async (observable) => {
          await firstValueFrom(
            observable.pipe(filter((result) => result.loading === false))
          );

          recentlyCreatedObservableQueries.delete(observable);
        })
      ).then(process);
    });
  }

  return Promise.resolve()
    .then(process)
    .finally(() => {
      availableObservableQueries.clear();
      recentlyCreatedObservableQueries.clear();
    });
}
