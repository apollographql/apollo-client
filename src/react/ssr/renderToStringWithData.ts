import { ReactElement } from 'react';
import { getMarkupFromTree } from './getDataFromTree';

export function renderToStringWithData(
  component: ReactElement<any>
): Promise<string> {
  return getMarkupFromTree({
    tree: component,
    renderFunction: require('react-dom/server').renderToString
  });
}
