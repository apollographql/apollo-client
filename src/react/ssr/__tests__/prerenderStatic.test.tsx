/* eslint-disable testing-library/render-result-naming-convention */
// not exported
// eslint-disable-next-line local-rules/no-relative-imports
import "../../../testing/internal/messageChannelPolyfill.js";

import "./polyfillReactDomTypes.d.ts";

import { expectTypeOf } from "expect-type";
import { JSDOM } from "jsdom";
import jsesc from "jsesc";
import * as React from "react";
import { renderToStaticMarkup, renderToString } from "react-dom/server";
import { prerender } from "react-dom/static";
import { prerenderToNodeStream } from "react-dom/static.node";

import type { TypedDocumentNode } from "@apollo/client";
import { ApolloClient, ApolloLink, gql, InMemoryCache } from "@apollo/client";
import {
  ApolloProvider,
  getApolloContext,
  useQuery,
  useSuspenseQuery,
} from "@apollo/client/react";
import { prerenderStatic } from "@apollo/client/react/ssr";
import { MockLink, MockSubscriptionLink } from "@apollo/client/testing";
import { resetApolloContext, wait } from "@apollo/client/testing/internal";
import { InvariantError } from "@apollo/client/utilities/invariant";

beforeEach(() => {
  // We are running tests with multiple different renderers, and that can result in a warning like
  // > Detected multiple renderers concurrently rendering the same context provider. This is currently unsupported.
  // This avoids that.
  resetApolloContext();
});
// @ts-ignore
global.setImmediate ||= (fn) => setTimeout(fn, 0);

function testSetup() {
  const query1: TypedDocumentNode<{ hello: string }> = gql`
    query {
      hello
    }
  `;
  const query2: TypedDocumentNode<{ whoami: { name: string } }> = gql`
    query {
      whoami {
        name
      }
    }
  `;
  const query3: TypedDocumentNode<{ currentTime: string }> = gql`
    query {
      currentTime
    }
  `;

  function Outlet() {
    const { data } = useSuspenseQuery(query1);
    return (
      <div>
        <p>Hello {data.hello}!</p>
        <Parent />
      </div>
    );
  }

  function Parent() {
    const { data } = useSuspenseQuery(query2);

    return (
      <>
        <p>My name is {data.whoami.name}!</p>
        <Child />
      </>
    );
  }
  function Child() {
    const { data } = useSuspenseQuery(query3);

    return (
      <>
        <p>Current time is {data.currentTime}!</p>
      </>
    );
  }

  const mocks = [
    {
      request: { query: query1 },
      result: { data: { hello: "world" } },
      maxUsageCount: 1,
    },
    {
      request: { query: query2 },
      result: { data: { whoami: { name: "Apollo" } } },
      maxUsageCount: 1,
    },
    {
      request: { query: query3 },
      result: { data: { currentTime: "2025-03-26T14:40:53.118Z" } },
      maxUsageCount: 1,
    },
  ];
  const mockLink = new MockLink(mocks);
  return {
    Outlet,
    mocks,
    mockLink,
    query1,
    query2,
    query3,
  };
}

test.each([
  ["prerender", prerender satisfies prerenderStatic.PrerenderToWebStream],
  [
    "prerenderToNodeStream",
    prerenderToNodeStream satisfies prerenderStatic.PrerenderToNodeStream,
  ],
] as const)(
  "real-life kitchen sink use case with %s and `useSuspenseQuery`",
  async (_, prerender) => {
    // reset things that might have been set by a previous run of this test with
    // a different runner
    (window as any).__APOLLO_CLIENT_INIT__ = undefined;
    window.document.documentElement.innerHTML = "<head></head><body></body>";

    const throwingLink = new ApolloLink(() => {
      throw new Error(
        "This ApolloClient instance should not need to make requests!"
      );
    });

    const { Outlet, mockLink: ssrLink } = testSetup();
    // from here on it's essentially what a userland app could look like

    function makeClient(link = throwingLink) {
      return new ApolloClient({
        cache: new InMemoryCache(),
        link,
      });
    }
    function App() {
      const externalClient = React.useContext(getApolloContext()).client;
      const [client] = React.useState<ApolloClient>(
        externalClient || makeClient
      );
      if ((window as any).__APOLLO_CLIENT_INIT__) {
        client.restore((window as any).__APOLLO_CLIENT_INIT__);
      }

      return (
        <html>
          <head>
            <meta charSet="utf-8" />
            <meta
              name="viewport"
              content="width=device-width, initial-scale=1"
            />
            <link rel="stylesheet" href="/styles.css"></link>
            <title>My app</title>
          </head>
          <body>
            <ApolloProvider client={client}>
              <React.Suspense fallback="Loading...">
                <Outlet />
              </React.Suspense>
            </ApolloProvider>
          </body>
        </html>
      );
    }

    async function ssr() {
      const client = makeClient(ssrLink);
      const signal = AbortSignal.timeout(2000);

      await prerenderStatic({
        tree: <App />,
        renderFunction: (tree) => prerender(tree, { signal }),
        context: {
          client,
        },
        ignoreResults: true,
        signal,
      });

      const extracted = client.extract();

      const { prelude } = await prerender(
        <ApolloProvider client={client}>
          <App />
        </ApolloProvider>,
        {
          bootstrapScripts: ["/main.js"],
          bootstrapScriptContent: `window.__APOLLO_CLIENT_INIT__ = ${jsesc(
            extracted,
            {
              isScriptContext: true,
              wrap: true,
              json: true,
            }
          )}`,
          signal,
        }
      );
      return new Response(prelude as any, {
        headers: { "content-type": "text/html" },
      });
    }

    const response = await ssr();

    // We're switching renderer, so we need to reset the context or the value
    // will carry over from the SSR render to the "browser" render
    resetApolloContext();

    const responseText = await response.text();
    expect(responseText).toMatchSnapshot();

    // JS isn't executed if we just assign the html to the current JSDOM root,
    // so we create another JSDOM instance just to extract the `__APOLLO_CLIENT_INIT__`
    // value and the top-level elements innerHTML
    const jsdom = new JSDOM(responseText, { runScripts: "dangerously" });
    expect(jsdom.window.__APOLLO_CLIENT_INIT__).toMatchSnapshot();
    const __APOLLO_CLIENT_INIT__ = jsdom.window.__APOLLO_CLIENT_INIT__;
    const innerHtml = jsdom.window.document.documentElement.innerHTML;
    jsdom.window.close();

    // hydrate the app in the "jest/browser" JSDOM
    (window as any).__APOLLO_CLIENT_INIT__ = __APOLLO_CLIENT_INIT__;
    window.document.documentElement.innerHTML = innerHtml;
    const reactClient = await import("react-dom/client");
    const root = reactClient.hydrateRoot(window.document, <App />, {
      onCaughtError: console.error,
      onRecoverableError: console.error,
      onUncaughtError: console.error,
    });

    // wait a bit - hydration errors would be thrown here
    await new Promise((resolve) => setTimeout(resolve, 500));

    // no errors, hydration was successful, test is done
    root.unmount();
  }
);

test.each([
  ["renderToString", renderToString satisfies prerenderStatic.RenderToString],
  [
    "renderToStaticMarkup",
    renderToStaticMarkup satisfies prerenderStatic.RenderToString,
  ],
])(
  "`prerenderStatic` with `%s` and suspense hooks will error",
  async (_, renderFunction) => {
    const { Outlet, mockLink } = testSetup();

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link: mockLink,
    });

    const promise = prerenderStatic({
      tree: <Outlet />,
      context: { client },
      renderFunction,
    });

    await expect(promise).rejects.toEqual(
      new Error(
        "A component suspended while responding to synchronous input. This will cause the UI to be replaced with a loading indicator. To fix, updates that suspend should be wrapped with startTransition."
      )
    );
  }
);

test.each([
  ["prerender", prerender],
  ["prerenderToNodeStream", prerenderToNodeStream],
])(
  "%s: AbortSignal times out during render - React re-throws abort error",
  async (_, renderFunction) => {
    const { Outlet, query1, query2 } = testSetup();
    type DataFor<T> = T extends TypedDocumentNode<infer D, any> ? D : never;

    const onError = jest.fn();

    const link = new MockSubscriptionLink();

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link,
    });

    const controller = new AbortController();

    const promise = prerenderStatic({
      tree: <Outlet />,
      context: { client },
      renderFunction: (tree) =>
        renderFunction(tree, {
          signal: controller.signal,
          onError,
        }),
      signal: controller.signal,
    });

    link.simulateResult(
      {
        result: {
          data: {
            hello: "world",
          } satisfies DataFor<typeof query1>,
        },
      },
      true
    );

    await wait(10);

    link.simulateResult(
      {
        result: {
          data: {
            whoami: { name: "Apollo" },
          } satisfies DataFor<typeof query2>,
        },
      },
      true
    );

    await wait(10);
    controller.abort("AbortReason");

    await expect(promise).rejects.toEqual("AbortReason");
    expect(onError).toHaveBeenLastCalledWith("AbortReason", expect.anything());
  }
);

test.each([
  ["renderToString", renderToString],
  ["renderToStaticMarkup", renderToStaticMarkup],
])(
  "%s: AbortSignal times out during render - stops rerendering, returns partial result",
  async (_, renderFunction) => {
    const { query1, query2, query3 } = testSetup();
    type DataFor<T> = T extends TypedDocumentNode<infer D, any> ? D : never;

    const link = new MockSubscriptionLink();

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link,
    });

    const controller = new AbortController();

    function Component() {
      const { data, loading } = useQuery(query1);
      if (loading) return <div>loading...</div>;
      return (
        <div>
          <p>Hello {data?.hello}!</p>
          <Parent />
        </div>
      );
    }

    function Parent() {
      const { data, loading } = useQuery(query2);
      if (loading) return <div>loading...</div>;

      return (
        <>
          <p>My name is {data?.whoami.name}!</p>
          <Child />
        </>
      );
    }
    function Child() {
      const { data, loading } = useQuery(query3);
      if (loading) return <div>loading...</div>;

      return (
        <>
          <p>Current time is {data?.currentTime}!</p>
        </>
      );
    }

    const promise = prerenderStatic({
      tree: <Component />,
      context: { client },
      renderFunction,
      signal: controller.signal,
      diagnostics: true,
    });

    link.simulateResult(
      {
        result: {
          data: {
            hello: "world",
          } satisfies DataFor<typeof query1>,
        },
      },
      true
    );

    await wait(10);

    link.simulateResult(
      {
        result: {
          data: {
            whoami: { name: "Apollo" },
          } satisfies DataFor<typeof query2>,
        },
      },
      true
    );

    await wait(10);

    controller.abort("AbortReason");
    // the network request here never "resolves"
    // (no call to `simulateResult` for the third query), yet the `abort` call
    // should finish up the prerendering

    const { result, aborted, diagnostics } = await promise;
    expect(result).toMatchSnapshot();
    expect(result).toMatch(/world/);
    expect(result).toMatch(/Apollo/);
    expect(result).toMatch(/loading.../);
    expect(aborted).toBe(true);
    expect(diagnostics?.renderCount).toBe(3);
  }
);

test("cancelled AbortSignal is passed into `prerenderStatic`", async () => {
  const { Outlet, mockLink } = testSetup();

  const controller = new AbortController();
  controller.abort();

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: mockLink,
  });

  const promise = prerenderStatic({
    tree: <Outlet />,
    context: { client },
    renderFunction: prerender,
    signal: controller.signal,
  });

  await expect(promise).rejects.toEqual(
    new Error("The operation was aborted before it could be attempted.")
  );
});

test("usage with `useSuspenseQuery`: `diagnostics.renderCount` stays 1", async () => {
  const { Outlet, mockLink } = testSetup();

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: mockLink,
  });

  const { diagnostics, result } = await prerenderStatic({
    tree: <Outlet />,
    context: { client },
    renderFunction: prerender,
    diagnostics: true,
  });

  expect(diagnostics?.renderCount).toBe(1);
  expect(result).toMatchInlineSnapshot(
    `"<div><p>Hello <!-- -->world<!-- -->!</p><p>My name is <!-- -->Apollo<!-- -->!</p><p>Current time is <!-- -->2025-03-26T14:40:53.118Z<!-- -->!</p></div>"`
  );
});

test("usage with `useQuery`: `diagnostics.renderCount` is 2", async () => {
  const { query1, mockLink } = testSetup();

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: mockLink,
  });

  function Component() {
    const { data, loading } = useQuery(query1);
    if (loading) return <div>loading...</div>;
    return <div>{data?.hello}</div>;
  }

  const { diagnostics, result } = await prerenderStatic({
    tree: <Component />,
    context: { client },
    renderFunction: prerender,
    diagnostics: true,
  });

  expect(diagnostics?.renderCount).toBe(2);
  expect(result).toMatchInlineSnapshot(`"<div>world</div>"`);
});

test("usage with a waterfall of `useQuery`: `diagnostics.renderCount` is `n+1`", async () => {
  const { query1, query2, query3, mockLink } = testSetup();

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: mockLink,
  });

  function Component() {
    const { data, loading } = useQuery(query1);
    if (loading) return <div>loading...</div>;
    return (
      <div>
        <p>Hello {data?.hello}!</p>
        <Parent />
      </div>
    );
  }

  function Parent() {
    const { data, loading } = useQuery(query2);
    if (loading) return <div>loading...</div>;

    return (
      <>
        <p>My name is {data?.whoami.name}!</p>
        <Child />
      </>
    );
  }
  function Child() {
    const { data, loading } = useQuery(query3);
    if (loading) return <div>loading...</div>;

    return (
      <>
        <p>Current time is {data?.currentTime}!</p>
      </>
    );
  }

  const { diagnostics, result } = await prerenderStatic({
    tree: <Component />,
    context: { client },
    renderFunction: prerender,
    diagnostics: true,
  });

  expect(diagnostics?.renderCount).toBe(4);
  expect(result).toMatchInlineSnapshot(
    `"<div><p>Hello <!-- -->world<!-- -->!</p><p>My name is <!-- -->Apollo<!-- -->!</p><p>Current time is <!-- -->2025-03-26T14:40:53.118Z<!-- -->!</p></div>"`
  );
});

test("multiple `useQuery` calls in the same component do not waterfall", async () => {
  const { query1, query2, query3, mockLink } = testSetup();

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: mockLink,
  });

  function Component() {
    const result1 = useQuery(query1);
    const result2 = useQuery(query2);
    const result3 = useQuery(query3);
    if (result1.loading || result2.loading || result3.loading)
      return <div>loading...</div>;

    return (
      <div>
        {result1.data?.hello}
        {result2.data?.whoami?.name}
        {result3.data?.currentTime}
      </div>
    );
  }

  const { result, diagnostics } = await prerenderStatic({
    tree: <Component />,
    context: { client },
    renderFunction: prerender,
    diagnostics: true,
  });

  expect(diagnostics?.renderCount).toBe(2);
  expect(result).toMatchInlineSnapshot(
    `"<div>world<!-- -->Apollo<!-- -->2025-03-26T14:40:53.118Z</div>"`
  );
});

test("`maxRerenders` will throw an error if exceeded", async () => {
  const query: TypedDocumentNode<{ hello: string }, { depth: number }> = gql`
    query ($depth: Int!) {
      hello(depth: $depth)
    }
  `;

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: new MockLink([
      {
        request: { query, variables: () => true },
        result: (arg) => ({ data: { hello: "world" + arg.depth } }),
        maxUsageCount: Number.POSITIVE_INFINITY,
      } satisfies MockLink.MockedResponse<{ hello: string }, { depth: number }>,
    ]),
  });

  function Component({ depth }: { depth: number }) {
    const { loading, data } = useQuery(query, { variables: { depth } });

    if (loading) return <div>loading...</div>;

    return (
      <div>
        {data?.hello}
        <Component depth={depth + 1} />
      </div>
    );
  }

  const promise = prerenderStatic({
    tree: <Component depth={1} />,
    context: { client },
    renderFunction: prerender,
    diagnostics: true,
    maxRerenders: 4,
  });
  await expect(promise).rejects.toEqual(
    new InvariantError(`Exceeded maximum rerender count of 4.
This either means you have very deep \`useQuery\` waterfalls in your application
and need to increase the \`maxRerender\` option to \`prerenderStatic\`, or that
you have an infinite render loop in your application.`)
  );
});

test("`maxRerenders` defaults to 50", async () => {
  const query: TypedDocumentNode<{ hello: string }, { depth: number }> = gql`
    query ($depth: Int!) {
      hello(depth: $depth)
    }
  `;

  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: new MockLink([
      {
        request: { query, variables: () => true },
        result: (arg) => ({ data: { hello: "world" + arg.depth } }),
        maxUsageCount: Number.POSITIVE_INFINITY,
      } satisfies MockLink.MockedResponse<{ hello: string }, { depth: number }>,
    ]),
  });

  function Component({ depth }: { depth: number }) {
    const { loading, data } = useQuery(query, { variables: { depth } });

    if (loading) return <div>loading...</div>;

    return (
      <div>
        {data?.hello}
        <Component depth={depth + 1} />
      </div>
    );
  }

  const promise = prerenderStatic({
    tree: <Component depth={1} />,
    context: { client },
    renderFunction: prerender,
    diagnostics: true,
  });
  await expect(promise).rejects.toEqual(
    new InvariantError(`Exceeded maximum rerender count of 50.
This either means you have very deep \`useQuery\` waterfalls in your application
and need to increase the \`maxRerender\` option to \`prerenderStatic\`, or that
you have an infinite render loop in your application.`)
  );
}, 10000);

test.each([
  ["renderToString", renderToString],
  ["renderToStaticMarkup", renderToStaticMarkup],
  ["prerender", prerender],
  ["prerenderToNodeStream", prerenderToNodeStream],
])(
  "`ignoreResults` will result in an empty string to be returned (%s)",
  async (_, renderFunction) => {
    const { mockLink, query1 } = testSetup();
    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link: mockLink,
    });
    function Component() {
      const { data, loading } = useQuery(query1);
      if (loading) return <div>loading...</div>;
      return <div>{data?.hello}</div>;
    }
    const { result } = await prerenderStatic({
      tree: <Component />,
      context: { client },
      renderFunction,
      ignoreResults: true,
    });

    expect(result).toBe("");
  }
);

it.skip("type tests", async () => {
  expectTypeOf(
    await prerenderStatic({
      tree: <div />,
      renderFunction: renderToStaticMarkup,
    })
  ).toEqualTypeOf<{
    result: string;
    aborted: boolean;
    diagnostics?: { renderCount: number };
  }>();
  expectTypeOf(
    await prerenderStatic({
      tree: <div />,
      renderFunction: renderToString,
    })
  ).toEqualTypeOf<{
    result: string;
    aborted: boolean;
    diagnostics?: { renderCount: number };
  }>();
  if (React.version.startsWith("19")) {
    const { prerender, prerenderToNodeStream } =
      require("react-dom/static") as typeof import("react-dom/static");

    expectTypeOf(
      await prerenderStatic({
        tree: <div />,
        renderFunction: prerender,
      })
    ).toEqualTypeOf<{
      result: string;
      aborted: boolean;
      diagnostics?: { renderCount: number };
    }>();
    expectTypeOf(
      await prerenderStatic({
        tree: <div />,
        renderFunction: prerenderToNodeStream,
      })
    ).toEqualTypeOf<{
      result: string;
      aborted: boolean;
      diagnostics?: { renderCount: number };
    }>();
  }
});
