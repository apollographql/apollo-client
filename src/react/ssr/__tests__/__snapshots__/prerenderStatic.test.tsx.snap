// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`real-life kitchen sink use case with prerender and \`useSuspenseQuery\` 1`] = `"<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\"/><link rel=\\"preload\\" as=\\"script\\" fetchPriority=\\"low\\" href=\\"/main.js\\"/><title>My app</title><link rel=\\"stylesheet\\" href=\\"/styles.css\\"/></head><body><!--$--><div><p>Hello <!-- -->world<!-- -->!</p><p>My name is <!-- -->Apollo<!-- -->!</p><p>Current time is <!-- -->2025-03-26T14:40:53.118Z<!-- -->!</p></div><!--/$--><script>window.__APOLLO_CLIENT_INIT__ = {\\"ROOT_QUERY\\":{\\"__typename\\":\\"Query\\",\\"hello\\":\\"world\\",\\"whoami\\":{\\"name\\":\\"Apollo\\"},\\"currentTime\\":\\"2025-03-26T14:40:53.118Z\\"}}</script><script src=\\"/main.js\\" async=\\"\\"></script></body></html>"`;

exports[`real-life kitchen sink use case with prerender and \`useSuspenseQuery\` 2`] = `
Object {
  "ROOT_QUERY": Object {
    "__typename": "Query",
    "currentTime": "2025-03-26T14:40:53.118Z",
    "hello": "world",
    "whoami": Object {
      "name": "Apollo",
    },
  },
}
`;

exports[`real-life kitchen sink use case with prerenderToNodeStream and \`useSuspenseQuery\` 1`] = `"<!DOCTYPE html><html><head><meta charSet=\\"utf-8\\"/><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\"/><link rel=\\"preload\\" as=\\"script\\" fetchPriority=\\"low\\" href=\\"/main.js\\"/><title>My app</title><link rel=\\"stylesheet\\" href=\\"/styles.css\\"/></head><body><!--$--><div><p>Hello <!-- -->world<!-- -->!</p><p>My name is <!-- -->Apollo<!-- -->!</p><p>Current time is <!-- -->2025-03-26T14:40:53.118Z<!-- -->!</p></div><!--/$--><script>window.__APOLLO_CLIENT_INIT__ = {\\"ROOT_QUERY\\":{\\"__typename\\":\\"Query\\",\\"hello\\":\\"world\\",\\"whoami\\":{\\"name\\":\\"Apollo\\"},\\"currentTime\\":\\"2025-03-26T14:40:53.118Z\\"}}</script><script src=\\"/main.js\\" async=\\"\\"></script></body></html>"`;

exports[`real-life kitchen sink use case with prerenderToNodeStream and \`useSuspenseQuery\` 2`] = `
Object {
  "ROOT_QUERY": Object {
    "__typename": "Query",
    "currentTime": "2025-03-26T14:40:53.118Z",
    "hello": "world",
    "whoami": Object {
      "name": "Apollo",
    },
  },
}
`;

exports[`renderToStaticMarkup: AbortSignal times out during render - stops rerendering, returns partial result 1`] = `"<div><p>Hello world!</p><p>My name is Apollo!</p><div>loading...</div></div>"`;

exports[`renderToString: AbortSignal times out during render - stops rerendering, returns partial result 1`] = `"<div><p>Hello <!-- -->world<!-- -->!</p><p>My name is <!-- -->Apollo<!-- -->!</p><div>loading...</div></div>"`;
